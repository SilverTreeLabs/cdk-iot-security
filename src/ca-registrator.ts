import * as path from 'path';
import {
  Role, PolicyStatement, Effect,
  ServicePrincipal, /*PolicyDocument,*/ ManagedPolicy,
} from '@aws-cdk/aws-iam';
import { Function } from '@aws-cdk/aws-lambda';
import { NodejsFunction } from '@aws-cdk/aws-lambda-nodejs';
import { Bucket } from '@aws-cdk/aws-s3';
import { Construct, Duration } from '@aws-cdk/core';

export interface CaRegistratorProps {
  activatorFunction: Function;
  activatorRole: Role;
  activatorQueueUrl: string;
  upload: UploadProps;
  verifiers?: [VerifierProps];
}

export interface UploadProps {
  bucket: Bucket;
  prefix: string;
  key: string;
}

export interface VerifierProps {
  name: string;
  lambdaFunction: Function;
}

export class CaRegistrator extends NodejsFunction {
  /**
   * Initialize the CA Registrator Function.
   * @param scope
   * @param id
   * @param props
   */
  constructor(scope: Construct, id: string, props: CaRegistratorProps) {
    let environment: {[key: string]: string} = {
      ACTIVATOR_ARN: props.activatorFunction.functionArn,
      ACTIVATOR_ROLE_ARN: props.activatorRole.roleArn,
      ACTIVATOR_QUEUE_URL: props.activatorQueueUrl,
      BUCKET_NAME: props.upload.bucket.bucketName,
      BUCKET_PREFIX: props.upload.prefix,
      BUCKET_KEY: props.upload.key,
    };
    props.verifiers?.forEach(verifier => environment[verifier.name] = verifier.lambdaFunction.functionArn);

    super(scope, `CaRegistratorFunction-${id}`, {
      entry: path.resolve(__dirname, './lambda-assets/caRegistrator/index.js'),
      role: new CaRegistationRole(scope, id),
      timeout: Duration.seconds(10),
      memorySize: 256,
      environment: environment,
    });
  }
}

class CaRegistationRole extends Role {
  /**
   * Initialize the CA Registration Role granted the neccessary
   * permission to register CA and create IoT Rule.
   * @param scope
   * @param id
   */
  constructor(scope: Construct, id:string) {
    super(scope, `CaRegistrationRole-${id}`, {
      roleName: `CaRegistrationRoleName-${id}`,
      assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName(
          'service-role/AWSLambdaBasicExecutionRole'),
      ],
    });
    this.addToPolicy(new PolicyStatement({
      effect: Effect.ALLOW,
      actions: [
        'iam:PassRole',
        'iam:CreateRole',
        'iam:AttachRolePolicy',
        'iot:RegisterCACertificate',
        'iot:TagResource',
        'iot:GetRegistrationCode',
        'iot:CreateTopicRule',
      ],
      resources: ['*'],
    }));
  }
}