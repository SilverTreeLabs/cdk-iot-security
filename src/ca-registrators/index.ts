export { CaRegistrationFunction } from './ca-registrator';
export { JitpCaRegistrationFunction } from './jitp-ca-registrator';
export { JitrCaRegistrationFunction } from './jitr-ca-registrator';