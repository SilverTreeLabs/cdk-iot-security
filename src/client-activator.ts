import {
  Role, PolicyStatement, Effect,
  ServicePrincipal, PolicyDocument, ManagedPolicy,
} from '@aws-cdk/aws-iam';
import { SqsEventSource } from '@aws-cdk/aws-lambda-event-sources';
import { NodejsFunction } from '@aws-cdk/aws-lambda-nodejs';
import { Queue } from '@aws-cdk/aws-sqs';
import { Construct } from '@aws-cdk/core';

export class ClientActivator extends Construct {
  public function: ActivationFunction;
  public role: Role;
  public receptor: Receptor;
  constructor(scope: Construct, id: string) {
    super(scope, `ClientActivator-${id}`);
    this.receptor = new Receptor(this, id);
    const activationRole = new ActivationRole(this, id);
    this.receptor.grantConsumeMessages(activationRole);
    this.function = new ActivationFunction(this, id, {
      activationRole: activationRole,
    });
    this.function.addEventSource(new SqsEventSource(this.receptor));
    this.role = this.receptor.getPushRole('iot.amazonaws.com');
    this.receptor.grantSendMessages(this.role);
  }
}

class Receptor extends Queue {
  constructor(scope: Construct, id: string) {
    super(scope, `Receptor-${id}`, {});
  }

  public getPushRole(principalName: string) {
    return new Role(this, `PushRole-${this.node.id}`, {
      roleName: `PushRole-${this.node.id}`,
      assumedBy: new ServicePrincipal(principalName),
      inlinePolicies: {
        SqsPushPolicy: new PolicyDocument({
          statements: [new PolicyStatement({
            actions: [
              'sqs:SendMessageBatch',
              'sqs:SendMessage',
            ],
            resources: [
              this.queueArn,
            ],
          })],
        }),
      },
    });
  }
}

interface ActivationFunctionProps {
  activationRole: Role;
}

class ActivationFunction extends NodejsFunction {
  constructor(scope: Construct, id: string, props: ActivationFunctionProps) {
    console.log(`${process.env.APPS_PATH}/activator/index.js`);
    super(scope, `ActivatorFunction-${id}`, {
      entry: `${process.env.APPS_PATH}/activator/index.js`,
      role: props.activationRole,
    });
  }
}

class ActivationRole extends Role {
  constructor(scope: Construct, id:string) {
    super(scope, `ActivatorRole-${id}`, {
      roleName: `ActivatorRole-${id}`,
      assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName(
          'service-role/AWSLambdaBasicExecutionRole'),
      ],
      inlinePolicies: {
        CaRegistrationPolicy: new PolicyDocument({
          statements: [new PolicyStatement({
            effect: Effect.ALLOW,
            actions: [
              'iot:UpdateCertificate',
              'iot:DescribeCertificate',
              'lambda:InvokeFunction',
              'lambda:InvokeAsync',
            ],
            resources: ['*'],
          })],
        }),
      },
    });
  }
}