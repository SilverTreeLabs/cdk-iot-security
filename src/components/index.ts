export {
  CaRegistrator,
} from './ca-registrator';
export {
  DeviceActivator,
} from './device-activator';
export {
  DeviceCertificateGenerator,
} from './deivce-certificate-generator';
export {
  FleetGenerator,
} from './fleet-generator';
export {
  FleetProvisioningRole,
  RegistrationConfigRole,
  ProvisionRole,
} from './provision-role';
export {
  ReviewReceptor,
  ReviewAcceptionRole,
  JitrTopicRule,
} from './review-receptor';
export {
  Vault,
  VaultProps,
} from './vault';
export {
  VerifiersFetcher,
} from './verifiers-fetcher';
export {
  GreenGrassV2TokenExchangeRole,
} from './greengrass-v2';