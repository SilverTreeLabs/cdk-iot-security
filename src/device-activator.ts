import * as path from 'path';
import {
  Role,
  PolicyStatement,
  Effect,
  ServicePrincipal,
  PolicyDocument,
  Policy,
} from '@aws-cdk/aws-iam';
// import * as lambda from '@aws-cdk/aws-lambda';
import { SqsEventSource } from '@aws-cdk/aws-lambda-event-sources';
import { NodejsFunction } from '@aws-cdk/aws-lambda-nodejs';
import * as sqs from '@aws-cdk/aws-sqs';
import { Construct } from '@aws-cdk/core';

// export module DeviceActivator {
// export class DeviceActivator extends Construct {
export class DeviceActivator extends Construct {
  /**
   * The Device Activation Function.
   */
  public function: DeviceActivator.Function;

  /**
   * The AWS SQS Queue collecting the messages received from the IoT rules.
   */
  public queue: DeviceActivator.Queue;

  /**
   * Initialize the Device Activator.
   *
   * The Device Activator is mainly consist of three parts,
   * a Lambda Function providing the Activation functionality,
   * a Receptor which is a SQS Queue receiving the messages
   * from the CA-associated Iot Rules created by the Registrator,
   * and a Role allowing pushing to the Receptor for granting the
   * Iot Rule through the Registrator.
   *
   * @param scope
   * @param id
   */
  constructor(scope: Construct, id: string) {
    super(scope, `DeviceActivator-${id}`);
    this.queue = new DeviceActivator.Queue(this, id);
    this.function = new DeviceActivator.Function(this, id);
    this.queue.grantConsumeMessages(this.function);
    this.function.addEventSource(
      new SqsEventSource(this.queue, { batchSize: 1 }),
    );
  }
}

export module DeviceActivator {
  export class Function extends NodejsFunction {
    /**
     * Inistialize the Device Activator Function.
     * @param scope
     * @param id
     * @param props
     */
    constructor(scope: Construct, id: string) {
      super(scope, `DeviceActivatorFunction-${id}`, {
        // code: lambda.Code.fromAsset(path.resolve(__dirname, '../lambda-assets/device-activator')),
        entry: path.resolve(__dirname, '../lambda-assets/device-activator/app.ts'),
        // handler: 'app.handler',
        // runtime: lambda.Runtime.NODEJS_14_X,
      });
      this.role?.attachInlinePolicy(
        new Policy(this, `DeviceActivationFunctionPolicy-${id}`, {
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                'iot:UpdateCertificate',
                'iot:DescribeCertificate',
                'lambda:InvokeFunction',
                'lambda:InvokeAsync',
              ],
              resources: ['*'],
            }),
          ],
        }),
      );
    }
  }

  export class Queue extends sqs.Queue {
    public readonly pushingRole: Queue.PushingRole;
    /**
     * Initialize the SQS Queue receiving message from the CA-associated Iot Rules.
     * @param scope
     * @param id
     */
    constructor(scope: Construct, id: string) {
      super(scope, `DeviceActivatorQueue-${id}`, {});
      this.pushingRole = new Queue.PushingRole(this, 'iot.amazonaws.com');
    }
  }
  export module Queue {
    /**
     * The Role allowing pushing messages into a specific Device Activator Queue.
     */
    export class PushingRole extends Role {
      constructor(queue: Queue, principalName: string) {
        let id = queue.node.id.replace('DeviceActivatorQueue', 'DeviceActivatorQueuePushingRole');
        let roleName = queue.node.id.replace('DeviceActivatorQueue', 'DeviceActivatorQueuePushingRoleName');
        super(queue, id, {
          roleName: roleName,
          assumedBy: new ServicePrincipal(principalName),
          inlinePolicies: {
            SqsPushPolicy: new PolicyDocument({
              statements: [
                new PolicyStatement({
                  actions: [
                    'sqs:SendMessageBatch',
                    'sqs:SendMessage',
                  ],
                  resources: [
                    queue.queueArn,
                  ],
                }),
              ],
            }),
          },
        });
      }
    }
  }
}