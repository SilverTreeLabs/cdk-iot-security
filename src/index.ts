export * from './just-in-time-registration';
export * from './just-in-time-provision';