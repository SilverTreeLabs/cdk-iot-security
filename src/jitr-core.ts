import { Construct } from '@aws-cdk/core';
import { RestApi } from '@aws-cdk/aws-apigateway';

import { ClientActivator } from './client-activator';
import { CaRegistrator, VerifierProps } from './ca-registrator';

export interface JitrCoreProps {
    verifiers?: [VerifierProps];
    api?: RestApi;
}

export class JitrCore extends Construct {
    public api: RestApi;
    public activator: ClientActivator;
    public registrator: CaRegistrator;
    constructor(scope: Construct, id: string, props: JitrCoreProps) {
        super(scope, `JitrCore-${id}`);
        this.activator = new ClientActivator(this, id);

        this.api = props.api || new RestApi(this, id);
        const resource = this.api.root.addResource('register');

        this.registrator = new CaRegistrator(this, id, {
            activatorFunction: this.activator.function,
            activatorRole: this.activator.role,
            activatorQueueUrl: this.activator.queue.queueUrl,
            apiResource: resource,
            verifiers: props.verifiers
        })
    }
}