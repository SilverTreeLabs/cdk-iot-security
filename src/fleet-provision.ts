import * as cdk from '@aws-cdk/core';
import { FleetGenerator } from './components/fleet-generator';
import { GreenGrassV2TokenExchangeRole } from './components/greengrass-v2';
import { FleetProvisionRole } from './components/provision-role';
import { VaultProps } from './components/vault';

export class FleetProvision extends cdk.Construct {
  public readonly fleetProvisionRole: FleetProvisionRole;
  public readonly fleetGenerator: FleetGenerator;
  public readonly greengrassV2TokenExchangeRole?: GreenGrassV2TokenExchangeRole;
  constructor(scope: cdk.Construct, id: string, props: FleetProvision.Props) {
    super(scope, id);
    if (props.greengrassV2) {
      this.greengrassV2TokenExchangeRole = new GreenGrassV2TokenExchangeRole(this, id);
    }
    this.fleetProvisionRole = new FleetProvisionRole(this, id);
    this.fleetGenerator = new FleetGenerator(this, id, {
      vault: props.vault,
      fleetProvisionRole: this.fleetProvisionRole,
      greengrassV2TokenExchangeRole: this.greengrassV2TokenExchangeRole,
    });
  }
}

export module FleetProvision {
  export interface Props {
    /**
     * The secure AWS S3 Bucket recepting the CA registration
     * information returned from the CA Registration Function.
     */
    readonly vault: VaultProps;
    readonly greengrassV2?: boolean;
  }
}