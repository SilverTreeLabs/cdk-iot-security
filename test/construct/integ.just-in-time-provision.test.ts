import * as path from 'path';
import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { Bucket } from '@aws-cdk/aws-s3';
import { App, Stack } from '@aws-cdk/core';
import { JustInTimeProvision } from '../../src';

describe('Integration test', () => {
  test('JustInTimeProvision', () => {
    process.env.BASE_PATH = __dirname;
    process.env.APPS_PATH = path.resolve(__dirname, '..', '..', 'src', 'lambda-assets');
    const app = new App();
    const stack = new Stack(app, 'test-stack');
    const name = 'test-case';
    const anotherStack = new Stack(app, 'another-stack');
    const caBucket = new Bucket(anotherStack, 'caBucket');
    const deviceVault = new Bucket(anotherStack, 'deviceBucket');
    new JustInTimeProvision(stack, name, {
      vault: {
        bucket: caBucket,
        prefix: 'test',
      },
      deviceVault: {
        bucket: deviceVault,
      },
    });
    expect(SynthUtils.synthesize(stack).template).toMatchSnapshot();
    expect(stack).toCountResources('AWS::Lambda::Function', 3);
    expect(stack).toCountResources('AWS::IAM::Role', 4);
  });
});