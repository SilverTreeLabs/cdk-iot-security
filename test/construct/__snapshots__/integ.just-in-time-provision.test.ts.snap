// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration test JustInTimeProvision 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters521c16efa99f6b226a4b5496c6851fc91fb6006c897e8cc62cc021278e3d4d71ArtifactHash3D5D3872": Object {
      "Description": "Artifact hash for asset \\"521c16efa99f6b226a4b5496c6851fc91fb6006c897e8cc62cc021278e3d4d71\\"",
      "Type": "String",
    },
    "AssetParameters521c16efa99f6b226a4b5496c6851fc91fb6006c897e8cc62cc021278e3d4d71S3Bucket736A0D41": Object {
      "Description": "S3 bucket for asset \\"521c16efa99f6b226a4b5496c6851fc91fb6006c897e8cc62cc021278e3d4d71\\"",
      "Type": "String",
    },
    "AssetParameters521c16efa99f6b226a4b5496c6851fc91fb6006c897e8cc62cc021278e3d4d71S3VersionKey2C9F0EDB": Object {
      "Description": "S3 key for asset version \\"521c16efa99f6b226a4b5496c6851fc91fb6006c897e8cc62cc021278e3d4d71\\"",
      "Type": "String",
    },
    "AssetParameters8282ef9000361ed699baec2f49978de2013260d755c9b26e4f92589fdd2cb319ArtifactHashCE7BC5EF": Object {
      "Description": "Artifact hash for asset \\"8282ef9000361ed699baec2f49978de2013260d755c9b26e4f92589fdd2cb319\\"",
      "Type": "String",
    },
    "AssetParameters8282ef9000361ed699baec2f49978de2013260d755c9b26e4f92589fdd2cb319S3Bucket349CDF32": Object {
      "Description": "S3 bucket for asset \\"8282ef9000361ed699baec2f49978de2013260d755c9b26e4f92589fdd2cb319\\"",
      "Type": "String",
    },
    "AssetParameters8282ef9000361ed699baec2f49978de2013260d755c9b26e4f92589fdd2cb319S3VersionKey6D764C38": Object {
      "Description": "S3 key for asset version \\"8282ef9000361ed699baec2f49978de2013260d755c9b26e4f92589fdd2cb319\\"",
      "Type": "String",
    },
    "AssetParameters8335c6298a97f0a0d18a8b5facfd149e57c1423f70cd31d9c34cc7d1ac0ffb7fArtifactHash6018291A": Object {
      "Description": "Artifact hash for asset \\"8335c6298a97f0a0d18a8b5facfd149e57c1423f70cd31d9c34cc7d1ac0ffb7f\\"",
      "Type": "String",
    },
    "AssetParameters8335c6298a97f0a0d18a8b5facfd149e57c1423f70cd31d9c34cc7d1ac0ffb7fS3BucketCA81C430": Object {
      "Description": "S3 bucket for asset \\"8335c6298a97f0a0d18a8b5facfd149e57c1423f70cd31d9c34cc7d1ac0ffb7f\\"",
      "Type": "String",
    },
    "AssetParameters8335c6298a97f0a0d18a8b5facfd149e57c1423f70cd31d9c34cc7d1ac0ffb7fS3VersionKey7F4F875D": Object {
      "Description": "S3 key for asset version \\"8335c6298a97f0a0d18a8b5facfd149e57c1423f70cd31d9c34cc7d1ac0ffb7f\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "JustInTimeProvisiontestcaseCaRegistratortestcase21FFC435": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "iam:PassRole",
                "iot:RegisterCACertificate",
                "iot:GetRegistrationCode",
                "iot:CreateTopicRule",
                "iot:TagResource",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "JustInTimeProvisiontestcaseCaRegistratortestcase21FFC435",
        "Roles": Array [
          Object {
            "Ref": "JustInTimeProvisiontestcaseCaRegistratortestcaseServiceRoleFB1CB993",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "JustInTimeProvisiontestcaseCaRegistratortestcase93903A27": Object {
      "DependsOn": Array [
        "JustInTimeProvisiontestcaseCaRegistratortestcaseServiceRoleDefaultPolicy8C8AE893",
        "JustInTimeProvisiontestcaseCaRegistratortestcaseServiceRoleFB1CB993",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters8282ef9000361ed699baec2f49978de2013260d755c9b26e4f92589fdd2cb319S3Bucket349CDF32",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters8282ef9000361ed699baec2f49978de2013260d755c9b26e4f92589fdd2cb319S3VersionKey6D764C38",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters8282ef9000361ed699baec2f49978de2013260d755c9b26e4f92589fdd2cb319S3VersionKey6D764C38",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "BUCKET_NAME": Object {
              "Fn::ImportValue": "another-stack:ExportsOutputRefcaBucketD1A50B2B031F53FA",
            },
            "BUCKET_PREFIX": "test",
            "REGISTRATION_CONFIG_ROLE_ARN": Object {
              "Fn::GetAtt": Array [
                "JustInTimeProvisiontestcaseProvisionRoleRegistrationConfigRoletestcaseE8C566C2",
                "Arn",
              ],
            },
            "VERIFIERS": "\\"[]\\"",
          },
        },
        "Handler": "index.handler",
        "MemorySize": 256,
        "Role": Object {
          "Fn::GetAtt": Array [
            "JustInTimeProvisiontestcaseCaRegistratortestcaseServiceRoleFB1CB993",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 10,
      },
      "Type": "AWS::Lambda::Function",
    },
    "JustInTimeProvisiontestcaseCaRegistratortestcaseServiceRoleDefaultPolicy8C8AE893": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::ImportValue": "another-stack:ExportsOutputFnGetAttcaBucketD1A50B2BArn68FC0213",
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::ImportValue": "another-stack:ExportsOutputFnGetAttcaBucketD1A50B2BArn68FC0213",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "JustInTimeProvisiontestcaseCaRegistratortestcaseServiceRoleDefaultPolicy8C8AE893",
        "Roles": Array [
          Object {
            "Ref": "JustInTimeProvisiontestcaseCaRegistratortestcaseServiceRoleFB1CB993",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "JustInTimeProvisiontestcaseCaRegistratortestcaseServiceRoleFB1CB993": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "JustInTimeProvisiontestcaseDeviceCertificateGeneratortestcase625C5945": Object {
      "DependsOn": Array [
        "JustInTimeProvisiontestcaseDeviceCertificateGeneratortestcaseServiceRoleDefaultPolicyB8C16E44",
        "JustInTimeProvisiontestcaseDeviceCertificateGeneratortestcaseServiceRoleCA89F5E4",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters8335c6298a97f0a0d18a8b5facfd149e57c1423f70cd31d9c34cc7d1ac0ffb7fS3BucketCA81C430",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters8335c6298a97f0a0d18a8b5facfd149e57c1423f70cd31d9c34cc7d1ac0ffb7fS3VersionKey7F4F875D",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters8335c6298a97f0a0d18a8b5facfd149e57c1423f70cd31d9c34cc7d1ac0ffb7fS3VersionKey7F4F875D",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "BUCKET_NAME": Object {
              "Fn::ImportValue": "another-stack:ExportsOutputRefcaBucketD1A50B2B031F53FA",
            },
            "BUCKET_PREFIX": "test",
          },
        },
        "Handler": "index.handler",
        "MemorySize": 256,
        "Role": Object {
          "Fn::GetAtt": Array [
            "JustInTimeProvisiontestcaseDeviceCertificateGeneratortestcaseServiceRoleCA89F5E4",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 10,
      },
      "Type": "AWS::Lambda::Function",
    },
    "JustInTimeProvisiontestcaseDeviceCertificateGeneratortestcasePolicyDeviceCertificateGeneratortestcaseD0261C2C": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "iot:DescribeThing",
                "iot:DeleteCertificate",
                "iot:DeleteThing",
                "iot:DescribeCACertificate",
                "iot:ListTagsForResource",
                "iot:DescribeEndpoint",
                "lambda:InvokeFunction",
                "lambda:InvokeAsync",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "JustInTimeProvisiontestcaseDeviceCertificateGeneratortestcasePolicyDeviceCertificateGeneratortestcaseD0261C2C",
        "Roles": Array [
          Object {
            "Ref": "JustInTimeProvisiontestcaseDeviceCertificateGeneratortestcaseServiceRoleCA89F5E4",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "JustInTimeProvisiontestcaseDeviceCertificateGeneratortestcaseServiceRoleCA89F5E4": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "JustInTimeProvisiontestcaseDeviceCertificateGeneratortestcaseServiceRoleDefaultPolicyB8C16E44": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::ImportValue": "another-stack:ExportsOutputFnGetAttcaBucketD1A50B2BArn68FC0213",
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::ImportValue": "another-stack:ExportsOutputFnGetAttcaBucketD1A50B2BArn68FC0213",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "JustInTimeProvisiontestcaseDeviceCertificateGeneratortestcaseServiceRoleDefaultPolicyB8C16E44",
        "Roles": Array [
          Object {
            "Ref": "JustInTimeProvisiontestcaseDeviceCertificateGeneratortestcaseServiceRoleCA89F5E4",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "JustInTimeProvisiontestcaseProvisionRoleRegistrationConfigRoletestcaseE8C566C2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "iot.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSIoTThingsRegistration",
              ],
            ],
          },
        ],
        "RoleName": "ProvisionRoleName-RegistrationConfigRole-test-case",
      },
      "Type": "AWS::IAM::Role",
    },
    "JustInTimeProvisiontestcaseVerifiersFetchertestcase669A84BA": Object {
      "DependsOn": Array [
        "JustInTimeProvisiontestcaseVerifiersFetchertestcaseServiceRole9B13C41F",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters521c16efa99f6b226a4b5496c6851fc91fb6006c897e8cc62cc021278e3d4d71S3Bucket736A0D41",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters521c16efa99f6b226a4b5496c6851fc91fb6006c897e8cc62cc021278e3d4d71S3VersionKey2C9F0EDB",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters521c16efa99f6b226a4b5496c6851fc91fb6006c897e8cc62cc021278e3d4d71S3VersionKey2C9F0EDB",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "VERIFIERS": "\\"[]\\"",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "JustInTimeProvisiontestcaseVerifiersFetchertestcaseServiceRole9B13C41F",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "JustInTimeProvisiontestcaseVerifiersFetchertestcaseServiceRole9B13C41F": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
